File: leetcode/python/000104.maximum-depth-of-binary-tree.py
from collections import deque
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # BFS
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0

        queue = deque([root])
        lvl = 0
        while queue:
            lvl += 1

            for _ in range(len(queue)):
                node = queue.popleft()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

        return lvl

    # Recursive DFS
    def maxDepth2(self, root: Optional[TreeNode]) -> int:
        def dfs(node, curr_depth):
            if node is None:
                return curr_depth

            return max(dfs(node.left, curr_depth + 1), dfs(node.right, curr_depth + 1))

        return dfs(root, 0)

    # Recursive DFS
    def maxDepth3(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0

        return 1 + max(self.maxDepth3(root.left), self.maxDepth3(root.right))

    # Iterative DFS
    def maxDepth4(self, root: Optional[TreeNode]) -> int:
        max_depth = 0
        stack = [(root, 1)]
        while stack:
            node, depth = stack.pop()

            if node:
                max_depth = max(max_depth, depth)
                stack.append((node.left, depth + 1))
                stack.append((node.right, depth + 1))

        return max_depth

File: leetcode/typescript/000104.maximum-depth-of-binary-tree.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

// Recursive DFS
function maxDepth(root: TreeNode | null): number {
  if (!root) {
    return 0;
  }

  return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
}
// Time complexity: O(n)
// Space complexity: O(n)

// Iterative BFS
class TreeNodeQueue {
  private queue: TreeNode[] = [];

  constructor(item?: TreeNode) {
    if (item) {
      this.queue.push(item);
    }
  }

  enqueue(item: TreeNode) {
    this.queue.push(item);
  }

  dequeue() {
    if (this.isEmpty()) {
      throw new Error("Queue is empty");
    }

    return this.queue.shift() as TreeNode;
  }

  isEmpty() {
    return this.queue.length === 0;
  }

  get length() {
    return this.queue.length;
  }
}

function maxDepthBFS(root: TreeNode | null): number {
  if (!root) {
    return 0;
  }

  let level = 0;
  const queue = new TreeNodeQueue(root);

  while (!queue.isEmpty()) {
    const levelSize = queue.length;

    // Empty current queue
    for (let i = 0; i < levelSize; i++) {
      const cur = queue.dequeue();

      // Add all it's children back to the queue
      if (cur.left) {
        queue.enqueue(cur.left);
      }

      if (cur.right) {
        queue.enqueue(cur.right);
      }
    }

    level++;
  }

  return level;
}
// Time complexity: O(n)
// Space complexity: O(n)

// Iterative DFS
function maxDepthDFSIterative(root: TreeNode | null): number {
  const stack: Array<[TreeNode | null, number]> = [[root, 1]];
  let res = 1;

  while (stack.length > 0) {
    const [node, level] = stack.pop()!;

    if (node) {
      res = Math.max(res, level);
      stack.push([node.left, level + 1]);
      stack.push([node.right, level + 1]);
    }
  }

  return res;
}
// Time complexity: O(n)
// Space complexity: O(n)
